athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseTableNameUtils.java

```diff
@@ -0,0 +1,152 @@
+    }
+
+    /**
+     * Helper which goes from an Athena Federation SDK TableName to an HBase TableName.
+     *
+     * @param tableName An Athena Federation SDK TableName.
+     * @return The corresponding HBase TableName.
+     */
+    public static org.apache.hadoop.hbase.TableName getQualifiedTable(TableName tableName)
+    {
+        return org.apache.hadoop.hbase.TableName.valueOf(getQualifiedTableName(tableName));
+    }
+
+    /**
+     * Gets the hbase table name from Athena table name. This is to allow athena to query uppercase table names
+     * (since athena does not support them). If an hbase table name is found with the athena table name, it is returned.
+     * Otherwise, tryCaseInsensitiveSearch is used to find the corresponding hbase table.
+     *
+     * @param tableName the case insensitive table name
+     * @return the hbase table name
+     */
+    public static org.apache.hadoop.hbase.TableName getHbaseTableName(HBaseConnection conn, TableName athTableName)
+            throws IOException
+    {
+        if (conn.tableExists(getQualifiedTable(athTableName))) {
+            return getQualifiedTable(athTableName);
+        }
+        Optional<org.apache.hadoop.hbase.TableName> caseInsensitiveMatch = tryCaseInsensitiveSearch(conn, athTableName);
+        if (caseInsensitiveMatch.isPresent()) {
+            return caseInsensitiveMatch.get();
+        }
+        else {
+            return null;
+        }
+    }
+
+    /**
+     * Performs a case insensitive table search by listing all table names in the schema (namespace), mapping them
+     * to their lowercase transformation, and then mapping the given tableName back to a unique table. To prevent ambiguity,
+     * an IllegalStateException is thrown if multiple tables map to the given tableName.
+     * @param conn the HBaseConnection used to retrieve the tables
+     * @param tableName The Athena TableName to find the mapping to
+     * @return The HBase TableName containing the found HBase table and the Athena Schema (namespace)
+     * @throws IOException 
+     */
+    @VisibleForTesting
+    protected static Optional<org.apache.hadoop.hbase.TableName> tryCaseInsensitiveSearch(HBaseConnection conn, TableName tableName)
+            throws IOException
+    {
+        logger.info("Table {} not found.  Falling back to case insensitive search.", tableName.getTableName());
+        Multimap<String, String> lowerCaseNameMapping = ArrayListMultimap.create();
+        org.apache.hadoop.hbase.TableName[] tableNames = conn.listTableNamesByNamespace(tableName.getSchemaName());
+        for (org.apache.hadoop.hbase.TableName nextTableName : tableNames) {
+            lowerCaseNameMapping.put(nextTableName.getQualifierAsString().toLowerCase(Locale.ENGLISH), nextTableName.getNameAsString());
+        }
+        Collection<String> mappedNames = lowerCaseNameMapping.get(tableName.getTableName());
+        if (mappedNames.size() > 1) {
```

Comment:
this needs to be tested; I am worried that an exception will bubble up to the console when loading the schema for the database which contain two tables with the same string (different capitalization)

Reply:
This exception does cause the query to fail (it was tested). This is the intended behavior.